content/blog/2020-12-18-stp59-final.md

```diff
@@ -0,0 +1,62 @@
+
+One particularly interesting hueristic I was able to identify for the computation of `n` was related to header stacks. Header stacks are a data type in P4 consisting of an array and a pointer to an index in the array. Fortunately, extracting from the packet into a header stack has the side effect of incrementing the index pointer. Even better, the sizes of header stacks is a compile-time known constant used by the type system, and the dynamic semantics of incrementing the stack pointer past the stack size is clearly defined. Therefore, parser loops extracting into header stacks are guaranteed to run a compile-time known constant number of iterations before terminating. Loops which extract to header stacks every iteration can be unrolled to `n+1` copies of straight-line code, where `n` is the size of the stack, and the semantics ought to be identical (this clearly merits a Coq verification in the future).
+
+Hueristics for computing `n` statically require further study, and indeed should never be able to acheive completeness due to decidability concerns.  As for my current version, it is limited in capacity for completeness because it uses the raw AST output by the parser. While this AST is preferable because it provides a pretty printer back to P4 source code, I am unable to support the header stack hueristic. To do so, I will need to switch to the typed AST, fully decorated with typing information by the type checker, which will be possible once pretty printing is supported. This done, computing `n` for header stacks becomes trivial, whereas it is unnecessarily difficult in my current version.
+
+## Evaluation
+
+I use three metrics for evaluation: correctness of the code transformation and the soundness and completeness of the static analysis tools.
+
+For correctness, I begin with a minimal collection of toy p4 programs which I developed for the sake of test-driven development of loop unrolling. Residing in `petr4/unroll-test`, these are edge-case control flow programs which are primarily expected-fail for the static analysis. In addition to these, Petr4 conveniently includes a robust test harness which runs the front-end on the entirety of p4c's benchmark suite. Combined with some custom test cases from the front-end development, this runs over 100 realistic p4 programs, each on several realistc packets, all in under a few seconds. In order to acquire reasonable evidence of the correctness of my transformation, I merely had to insert a call to my unroller in this test harness before type checking. After some work, my tool achieves the same correctness as the Petr4's evaluater on unmodified code. The only limitation of this approach is that it does not include any negative tests, so I am much more sure of the correctness of my code transformation tool than I am of my static analysis tool.
```

Comment:
I again think that including some small code examples here could really help illustrate what's going on.